feat: support files listed in nonRenderableFiles with conditionalFiles

derberg opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow contributing guidelines and make sure PR title follows Conventional Commits specification -> https://github.com/asyncapi/generator/blob/master/CONTRIBUTING.md
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

**Description**

- nonRenderableFiles are also supported by conditionalFiles.
- extend some details in authoring.md

**Related issue(s)**
Resolves https://github.com/asyncapi/generator/issues/358

derberg commented:
@fmvilas 

I added support for nonrenderablefiles as agreed in the related issue

I tested against:
```
"nonRenderableFiles": [
      "src/api/middlewares/*.*",
      "src/lib/config.js"
    ],
    "conditionalFiles": {
      "src/lib/config.js": {
        "subject": "server.protocol",
        "validation": {
          "const": "amqp"
        }
      }
    }
```
and works as expected, `src/lib/config.js` is not getting through

ready for review

Tenischev commented:
@derberg i do not agree that this will resolve #358
First of all, example of "subject": "server.protocol", is not sufficient since there is no hint that every time when you wish to generate template, in the fact, you **must** specify `server` paramter otherwise `conditionalFiles` will no work.

derberg commented:
@Tenischev it is not yet merged, so for sure it doesn't resolve yet. But the goal is that it does 😉 

> First of all, example of "subject": "server.protocol", is not sufficient since there is no hint that every time when you wish to generate template, in the fact, you must specify server paramter otherwise conditionalFiles will no work.

- there are 2 examples now
- descriptions of server param is enriched

Do you like it 😄 
If not I need some hints on how to make you happy 😄

Tenischev commented:
I ❤️  new `server` parameter description 👏 😄 

But there still issue about these 
> if `subject` is not resolved in AsyncAPI, then file will be rendered because of [this condition](https://github.com/asyncapi/generator/blob/fc4dee9af622dfb198471a2cf0da53f5a3c7c7ae/lib/generator.js#L565)

derberg commented:
@Tenischev oh, I know what you mean, that actually sucks, you are right, in this new example I added about `info.contact`, if entire `contact` is not provided, `support.html` doc will be created

so I made a fix but to fix it nice there would have to be a nice log message. But we have too many of those already, so I added another use case to our new `debug` flag.

example of log messages for different cases, where there is debug on or off
```
$ ./cli.js ../nodejs-template/asyncapi.yml ../nodejs-template -o output --force-write -p server=dummy-mqtt --debug
docs/support.html was not generated because info.contact specified in template configuration in conditionalFiles was not found in provided AsyncAPI specification file

Done! ✨
Check out your shiny new generated files at /Users/wookiee/Documents/sources/generator/output.

$ ./cli.js ../nodejs-template/asyncapi.yml ../nodejs-template -o output --force-write -p server=dummy-mqtt --debug

docs/support.html was not generated because condition specified for this file in template configuration in conditionalFiles matched

Done! ✨
Check out your shiny new generated files at /Users/wookiee/Documents/sources/generator/output.

$ ./cli.js ../nodejs-template/asyncapi.yml ../nodejs-template -o output --force-write -p server=dummy-mqtt


Done! ✨
Check out your shiny new generated files at /Users/wookiee/Documents/sources/generator/output.
```

Tenischev commented on docs/authoring.md:
```diff
@@ -211,17 +211,17 @@ The `generator` property from `package.json` file must contain a JSON object tha
   },
   "conditionalFiles": {
-    "src/api/adapters/amqp.js": {
+    "template/dir/relative/path/to/file/test-amqp.js": {
```
Without `template` in the beginning, because of https://github.com/asyncapi/generator/blob/736bfa0108d158da07d86acaffbc57438c402f73/lib/generator.js#L546

Tenischev commented on docs/authoring.md:
```diff
@@ -211,17 +211,17 @@ The `generator` property from `package.json` file must contain a JSON object tha
-      "validation": {
-        "const": "mqtt"
-      }
+    "template/dir/relative/path/to/file/support.html": {
```
Without `template` in the beginning, because of https://github.com/asyncapi/generator/blob/736bfa0108d158da07d86acaffbc57438c402f73/lib/generator.js#L546

derberg replied:
ok, I think I made it a bit too confusing, I meant here in example that path must be relative to template dir. I'll fix

Tenischev commented on lib/generator.js:
```diff
@@ -158,6 +159,8 @@ class Generator {
       throw new Error('Parameter entrypoint is required when using output = "string"');
     }
 
+    if (this.debug) log.setLevel('debug');
```
Nice idea! Could we pass debug flag into template, so in template we also could write some debug logs. (Ofc this suggestion is for separate issue)

derberg replied:
you for sure have access to it in hooks out of the box, but filters and Nunjucks, that would be awesome, please write an issue for it

derberg commented:
@Tenischev how about https://github.com/asyncapi/generator/pull/363/commits/1c0dccd4774ff15506e58a731b9d7127beabb051
